substrate/frame/asset-rewards/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+				.or_else(|_| ensure_signed(origin))?;
+
+			let mut pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == caller, BadOrigin);
+			pool_info.admin = new_admin.clone();
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolAdminModified { pool_id, new_admin });
+
+			Ok(())
+		}
+
+		/// Modify a expiry block.
+		#[pallet::call_index(6)]
+		pub fn set_pool_expiry_block(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			new_expiry_block: BlockNumberFor<T>,
+		) -> DispatchResult {
+			let caller = T::PermissionedOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			ensure!(
+				new_expiry_block > frame_system::Pallet::<T>::block_number(),
+				Error::<T>::ExpiryBlockMustBeInTheFuture
+			);
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == caller, BadOrigin);
+
+			// Always start by updating the pool rewards.
+			let mut pool_info = Self::update_pool_rewards(&pool_info)?;
+
+			pool_info.expiry_block = new_expiry_block;
+			Pools::<T>::insert(pool_id, pool_info);
+
+			Self::deposit_event(Event::PoolExpiryBlockModified { pool_id, new_expiry_block });
+
+			Ok(())
+		}
+
+		/// Convinience method to deposit reward tokens into a pool.
+		///
+		/// This method is not strictly necessary (tokens could be transferred directly to the
+		/// pool pot address), but is provided for convenience so manual derivation of the
+		/// account id is not required.
+		#[pallet::call_index(7)]
+		pub fn deposit_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let caller = ensure_signed(origin)?;
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			let pool_account_id = Self::pool_account_id(&pool_id)?;
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&caller,
+				&pool_account_id,
+				amount,
+				Preservation::Preserve,
+			)?;
+			Ok(())
+		}
+
+		/// Permissioned method to withdraw reward tokens from a pool.
+		#[pallet::call_index(8)]
+		pub fn withdraw_reward_tokens(
+			origin: OriginFor<T>,
+			pool_id: PoolId,
+			amount: T::Balance,
+			dest: T::AccountId,
+		) -> DispatchResult {
+			let caller = T::PermissionedOrigin::ensure_origin(origin.clone())
+				.or_else(|_| ensure_signed(origin))?;
+
+			let pool_info = Pools::<T>::get(pool_id).ok_or(Error::<T>::NonExistentPool)?;
+			ensure!(pool_info.admin == caller, BadOrigin);
+
+			T::Assets::transfer(
+				pool_info.reward_asset_id,
+				&Self::pool_account_id(&pool_id)?,
+				&dest,
+				amount,
+				Preservation::Preserve,
+			)?;
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Derive a pool account ID from the pallet's ID.
+		pub fn pool_account_id(id: &PoolId) -> Result<T::AccountId, DispatchError> {
+			if Pools::<T>::contains_key(id) {
+				Ok(T::PalletId::get().into_sub_account_truncating(id))
```

Comment:
I think if you can get it to use the non-truncating version the it should be a bit safer - just to ensure that pool accounts dont overlap.

Reply:
Sure, I'll add an integrity test that ensures the AccountId is > 8 bytes.